### PR TITLE
KAFKA-17598; Command line validation tool for RE2J regex

### PR DESCRIPTION
**Pull request by dajac**
This patch introduces the `--validate-regex` argument to the `kafka-consumer-group` command line tool as defined in KIP-848. The new argument allows the verification of RE2 regular expressions.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java [hunk 1]
```diff
@@ -95,7 +97,14 @@ public static void main(String[] args) {
...
+            long actions = Stream.of(
+                opts.listOpt,
+                opts.describeOpt,
+                opts.deleteOpt,
+                opts.resetOffsetsOpt,
+                opts.deleteOffsetsOpt,
+                opts.validateRegexOpt
+            ).filter(opts.options::has).count();
             if (actions != 1)
                 CommandLineUtils.printUsageAndExit(opts.parser, "Command must include exactly one action: --list, --describe, --delete, --reset-offsets, --delete-offsets");
```
Comment (AndrewJSchofield) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java]: Probably ought to add `validate-regex` into the usage message.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]
lgtm apart from one small comment.

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java (hunk 1 above)
Comment (dajac) [reply tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java]: Nice catch! I fixed it. Result:
```
Command must include exactly one action: --list, --describe, --delete, --reset-offsets, --delete-offsets, --validate-regex
```

**Review by AndrewJSchofield — APPROVED** [review]

**Comment by AndrewJSchofield** [comment]
A little checkstyle to fix :)

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java [hunk 2]
```diff
@@ -130,6 +155,15 @@ else if (opts.options.has(opts.resetOffsetsOpt)) {
         }
     }
 
+    static void validateRegex(String regex) {
+        try {
+            Pattern.compile(regex);
+            System.out.printf("The regular expressions `%s` is valid.%n", regex);
```
Comment (lianetm) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java]:
```suggestion
            System.out.printf("The regular expression `%s` is valid.%n", regex);
```

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java [hunk 3]
```diff
@@ -130,6 +155,15 @@ else if (opts.options.has(opts.resetOffsetsOpt)) {
...
+        } catch (PatternSyntaxException ex) {
+            System.out.printf("The regular expressions `%s` is invalid: %s.%n", regex, ex.getDescription());
```
Comment (lianetm) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommand.java]:
```suggestion
            System.out.printf("The regular expression `%s` is invalid: %s.%n", regex, ex.getDescription());
```

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommandOptions.java [hunk 4]
```diff
@@ -210,7 +216,9 @@ private ConsumerGroupCommandOptions(String[] args) {
     void checkArgs() {
         CommandLineUtils.maybePrintHelpOrVersion(this, "This tool helps to list all consumer groups, describe a consumer group, delete consumer group info, or reset consumer group offsets.");
 
-        CommandLineUtils.checkRequiredArgs(parser, options, bootstrapServerOpt);
+        if (!options.has(validateRegexOpt)) {
+            CommandLineUtils.checkRequiredArgs(parser, options, bootstrapServerOpt);
+        }
```
Comment (lianetm) [inline tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommandOptions.java:221]: agree, but should we then update the bootstrap doc that states it's required?

**Review by lianetm — COMMENTED** [review]
Great addition! just a couple of nits

File: tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommandOptions.java (hunk 4 above)
Comment (dajac) [reply tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommandOptions.java:221]: yeah, i considered doing this but i did not like it. how would you state it? "required except for --validate-regex"?

Comment (lianetm) [reply tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommandOptions.java:221]: it currently is "REQUIRED: The server(s) to connect to.". Maybe flipping it to "The server(s) to connect to. REQUIRED for all options except for --validate-regex"?

Comment (dajac) [reply tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommandOptions.java:221]: works for me

Comment (dajac) [reply tools/src/main/java/org/apache/kafka/tools/consumer/group/ConsumerGroupCommandOptions.java:221]: fixed

**Review by lianetm — APPROVED** [review]
LGTM. THanks!

**Pull request merged**